test/gtest/ucp/test_ucp_am.cc

```diff
@@ -799,6 +800,192 @@ UCS_TEST_P(test_ucp_am_nbx, rx_am_mpools,
 UCP_INSTANTIATE_TEST_CASE(test_ucp_am_nbx)
 
 
+class test_ucp_am_nbx_send_copy_header : public test_ucp_am_nbx {
+private:
+    static const ucs_status_ptr_t send_completed;
+    size_t req_counter = 0;
+    size_t cb_counter = 0;
+    std::string m_hdr_copy;
+
+    ucs_status_ptr_t send_am(const ucp::data_type_desc_t &dt_desc,
+                            unsigned flags = 0, const void *hdr = NULL,
+                            unsigned hdr_length = 0,
+                            const ucp_mem_h memh = NULL) {
+        req_counter++;
+        return test_ucp_am_nbx::send_am(dt_desc, flags, hdr, hdr_length, memh);
+    }
+
+    ucs_status_ptr_t fill_sq(ucp::data_type_desc_t &sdt_desc, const ucp_mem_h memh, size_t header_size = 0ul,
+                           unsigned flags = 0, unsigned data_cb_flags = 0) {
+        double timeout = 5;
+        ucs_status_ptr_t pending_sptr = NULL;
+
+        //warmup for wireup connection
+        pending_sptr = send_am(sdt_desc, get_send_flag() | flags,
+                                            m_hdr_copy.data(), header_size, memh);
+        wait_for_flag(&m_am_received);
+        request_wait(pending_sptr);
+        pending_sptr = NULL;
+
+        ucs_time_t deadline = ucs_get_time() +
+                              ucs_time_from_sec(timeout) * ucs::test_time_multiplier();
+        while ((ucs_get_time() < deadline) && pending_sptr == send_completed) {
+            pending_sptr = send_am(sdt_desc, get_send_flag() | flags,
+                                            m_hdr_copy.data(), header_size, memh);
+        }
+
+        return pending_sptr;
+    }    
+protected:
+    void check_eager(size_t size, size_t header_size = 0ul,
+                           unsigned flags = 0, unsigned data_cb_flags = 0) {
+        mem_buffer sbuf(size, tx_memtype());
+        sbuf.pattern_fill(SEED);
+        m_hdr_copy.resize(header_size);
+        ucs::fill_random(m_hdr_copy);
+        m_hdr = m_hdr_copy;
+        ucs_status_ptr_t pending_sptr = NULL;
+        req_counter = 0;
+        cb_counter = 0;
+
+        ucp::data_type_desc_t sdt_desc(m_dt, sbuf.ptr(), size);                               
+        ucp_mem_h memh;
+
+        if (prereg()) {
+            memh = sender().mem_map(sbuf.ptr(), size);
+        }
+
+        set_am_data_handler(receiver(), TEST_AM_NBX_ID, am_data_cb, this,
+                                    data_cb_flags);
+
+        pending_sptr = fill_sq(sdt_desc, memh, header_size, get_send_flag() | flags, data_cb_flags);
+
+        if (pending_sptr == send_completed) {
+            UCS_TEST_MESSAGE <<"Failed to get pending request";
+        }
+
+        ucs::fill_random(m_hdr_copy);
+
+        while(progress());
+        wait_for_value(&cb_counter, req_counter);
+        request_wait(pending_sptr);
+        EXPECT_EQ(req_counter, cb_counter);
+
+        if (prereg()) {
+            sender().mem_unmap(memh);
+        }
+    }
+
+    void check_rndv(size_t size, size_t header_size = 0ul,
```

Comment:
Unresvoled

Reply:
Hi. Do you still think that this comment is relevant?
I use a helper function (fill_sq) to avoid code duplication. They looks similar but there are delicate differences between the two, delicate but important...